src/dump.rs

```diff
@@ -0,0 +1,281 @@
+                if let LeafOrSubleaves::Subleaf(leaves) = ent {
+                    leaves.remove(&subleaf);
+                } else {
+                    panic!("removing a subleaf when there's a leaf. no");
+                }
+            });
+        }
+
+        self.update_max_leaves();
+    }
+}
+
+impl CpuIdDump {
+    fn update_max_leaves(&mut self) {
+        let mut max_standard = None;
+        let mut max_hv = None;
+        let mut max_extended = None;
+
+        for k in self.leaves.keys() {
+            let k = *k;
+            if k < 0x40000000 {
+                max_standard = Some(match max_standard {
+                    None => k,
+                    Some(prev) => core::cmp::max(k, prev),
+                });
+            } else if k < 0x80000000 {
+                max_hv = Some(match max_hv {
+                    None => k,
+                    Some(prev) => core::cmp::max(k, prev),
+                });
+            } else if k < 0xc0000000 {
+                max_extended = Some(match max_extended {
+                    None => k,
+                    Some(prev) => core::cmp::max(k, prev),
+                });
+            }
+        }
+
+        if let Some(eax) = max_standard {
+            match self
+                .leaves
+                .entry(0)
+                .or_insert(LeafOrSubleaves::Leaf(CpuIdResult::empty()))
+            {
+                LeafOrSubleaves::Leaf(leaf) => {
+                    leaf.eax = eax;
+                }
+                _ => {
+                    panic!("cannot update leaf 1.EAX: leaf has subleaves?");
+                }
+            }
+        }
+
+        if let Some(eax) = max_hv {
+            match self
+                .leaves
+                .entry(0x40000000)
+                .or_insert(LeafOrSubleaves::Leaf(CpuIdResult::empty()))
+            {
+                LeafOrSubleaves::Leaf(leaf) => {
+                    leaf.eax = eax;
+                }
+                _ => {
+                    panic!("cannot update leaf 0x40000000.EAX: leaf has subleaves?");
+                }
+            }
+        }
+
+        if let Some(eax) = max_extended {
+            match self
+                .leaves
+                .entry(0x80000000)
+                .or_insert(LeafOrSubleaves::Leaf(CpuIdResult::empty()))
+            {
+                LeafOrSubleaves::Leaf(leaf) => {
+                    leaf.eax = eax;
+                }
+                _ => {
+                    panic!("cannot update leaf 0x80000000.EAX: leaf has subleaves?");
+                }
+            }
+        }
+    }
+}
+
+impl CpuIdReader for CpuIdDump {
+    fn cpuid1(&self, leaf: u32) -> CpuIdResult {
+        match self.leaves.get(&leaf) {
+            Some(LeafOrSubleaves::Leaf(res)) => *res,
+            Some(LeafOrSubleaves::Subleaf(subleaves)) => {
+                *subleaves.get(&0).unwrap_or_else(|| {
+                    // TODO: vendor-specific fallback behavior
```

Comment:
AMD and Intel have different behavior when querying leaves/subleaves that aren't actually defined and don't have behavior otherwise defined. I've fully punted here but would fill this out if you want the patch.